notes/trees-that-grow-proc-macro.md

```diff
@@ -0,0 +1,167 @@
+As nickel grows, it becomes unwieldy to use a single AST for every stage of compilation. Some things are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types for the LSP, closures at runtime). Other things are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.
+
+## Strategies
+
+One strategy to deal with this is to repeat a large AST enum in multiple places with small differences, and convert from one to another as we move through compilation. This creates a large maintenance burden, since many changes have to be propagated to all the trees.
```

Comment:
```suggestion
One strategy to deal with this is to repeat a large AST enum in multiple places with small differences, and convert from one to another as we move through program manipulations. This creates a large maintenance burden, since many changes have to be propagated to all the trees.
```